ecosystem/node-checker/vfn-check-client/src/big_query.rs

```diff
@@ -0,0 +1,191 @@
+    pub big_query_key_path: PathBuf,
+
+    /// GCP project ID.
+    #[clap(long, default_value = "analytics-test-345723")]
+    pub gcp_project_id: String,
+
+    /// BigQuery dataset ID.
+    #[clap(long, default_value = "ait3_vfn_nhc")]
+    pub big_query_dataset_id: String,
+
+    /// BigQuery table ID.
+    #[clap(long, default_value = "nhc_response_data")]
+    pub big_query_table_id: String,
+}
+
+// This struct formats the data into a format that BigQuery expects.
+#[derive(Debug, Serialize)]
+pub struct MyBigQueryRow {
+    pub account_address: String,
+    pub nhc_response_json: String,
+    pub time_response_received: u64,
+}
+
+impl TryFrom<(AccountAddress, SingleCheck)> for MyBigQueryRow {
+    type Error = anyhow::Error;
+
+    fn try_from(
+        (account_address, single_check): (AccountAddress, SingleCheck),
+    ) -> Result<Self, Self::Error> {
+        Ok(Self {
+            account_address: account_address.to_string(),
+            nhc_response_json: serde_json::to_string(&single_check.result)
+                .context("Failed to encode result data as JSON")?,
+            time_response_received: single_check.timestamp.as_secs(),
+        })
+    }
+}
+
+/// Instead of reading whether the dataset / table exists and then conditionally
+/// creating them, we just try every time and ignore the ALREADY_EXISTS error.
+/// This function does that check.
+fn ignore_already_exists_error<T>(result: Result<T, BQError>) -> Result<(), BQError> {
+    match result {
+        Ok(_) => Ok(()),
+        Err(err) => match err {
+            BQError::ResponseError { ref error } => {
+                if error.error.status == "ALREADY_EXISTS" {
+                    Ok(())
+                } else {
+                    Err(err)
+                }
+            }
+            wildcard => Err(wildcard),
+        },
+    }
+}
+
+/// Make a table field required, as part of the table creation request.
+fn make_required(mut table_field_schema: TableFieldSchema) -> TableFieldSchema {
+    table_field_schema.mode = Some("REQUIRED".to_string());
+    table_field_schema
+}
+
+/// Write the response data to BigQuery. Note: On the first run this fails
+/// sometimes because it doesn't seem BigQuery offers read-what-you-wrote.
+/// As in, it'll make the table, but then fail to insert the data because
+/// the table apparently doesn't exist, but then succeed the next time.
+pub async fn write_to_big_query(
+    big_query_args: &BigQueryArgs,
+    nhc_responses: HashMap<AccountAddress, Vec<SingleCheck>>,
+) -> Result<()> {
+    let client = BigQueryClient::from_service_account_key_file(
+        big_query_args
+            .big_query_key_path
+            .to_str()
+            .context("Big query key path was invalid")?,
+    )
+    .await;
+
+    info!(
+        "Creating dataset if necessary: {}",
+        big_query_args.big_query_dataset_id
+    );
+
+    // Create the dataset if necessary.
+    ignore_already_exists_error(
+        client
```

Comment:
nit: I'd maybe move each of the dataset creation calls into helpers 😄, e.g., the one below is pretty big.